RFC: HWM2: Extend to include SiP support

bjarki-andreasen opened the pull request:
## Introduction
We at Nordic have quite a few SiPs which contain the same SoC. A SiP is essentially an "integrated board", typically consisting of an SoC, surrounded by hardware like PMICs, external flash and clocks etc, with a subset of the SoCs connections routed to the pins of the SiP.

This PR makes it possible to separate SoCs and SiPs, defining the SoC in `soc/`, and SiP in `sip/`. This allows us to accurately add a SiP from one vendor, which contains one or more SoCs, potentially from different vendors with additional hardware around it, onto a board, and using the SiPs name as the qualifier when building for the board. The SoC is then selected by the SiP, along with devicetree overlays etc.

Summarized, after this PR (though not in its current state), the board qualifiers will look like
```
<board>/<sip>/<soc>/<cpucluster>/<variant>
```
if a sip is provided as the second qualifier, otherwise it falls back to what we already have
```
<board>/<soc>/<cpucluster>/<variant>
```
so its backwards compatible while providing a new neccesary layer to the HWM2. If the sip only has one soc:
```
<board>/<sip>//<cpucluster>/<variant>
```
and lastly, sip with only one soc with one cpucluster:
```
<board>/<sip>///<variant>
```

## What does it solve
SiPs like the nrf9161 and nrf9280 are not SoCs, but we currently have to treat them as such, adding an entry for them in `soc.yml`:
https://github.com/zephyrproject-rtos/zephyr/blob/f3bf6357c799191c230142ecb04994c0f232ed16/soc/nordic/soc.yml#L55
then adding a layer of misdirection using Kconfig to select the actual SoC:
https://github.com/zephyrproject-rtos/zephyr/blob/f3bf6357c799191c230142ecb04994c0f232ed16/soc/nordic/nrf91/Kconfig.soc#L39-L41
This solution only works until:
* Addressable hardware, like sensors, external RAM/flash etc. has been connected to the SoC within the SiP
* The SoC within the SiP is not from the same vendor, so can not be "quietly inserted" into the `soc.yml` as a completely 
  seperate SoC, and even if so, without essentially including the entire SoC vendor's SoC folder from the SiP vendor's "SoC" 
  folder. Consider having to add
  ```
  socs:
    - name: nrf9161
    - name: foo8312
  ```
  to `soc/nordic/soc.yml` and adding SiP specific options in `soc/foo/foo8312/`. Not nice...

## How does it work
SiPs are defined in a `sip.yml` file similar to SoCs, which looks like:
```
sips:
  - name: foo_sip
    socs:
      - bar_soc
```
The SoC is one defined in an `soc.yml` file, which contains the SoC's family, cpuclusters etc:
```
family:
  - name: bar_family
    series:
      - name: bar_series
        socs:
          - name: bar_soc
            cpuclusters:
              - name: cpuapp
              - name: cpunet
```
The sip can then be added to a `board.yml`, just like socs, with variants and all (though the variant requires the target soc along with cpucluster):
```
board:
  name: plank
  full_name: PLANK V0
  sips:
    - name: foo_sip
      variants:
        - name: 'ns'
          soc: bar_soc
          cpucluster: 'cpuapp'
```
When selecting the board for building, the sip is used as a qualifier in place of an soc:
```
plank/foo_sip/bar_soc/cpuapp/ns
```
The cpucluster is inherited from the soc contained in the sip as defined in sip.yml.
Lastly, the board selects the sip, rather than the soc, in its Kconfig.plank:
```
config BOARD_PLANK
        select SIP_FOO
```
SIP_FOO is defined in a Kconfig in `sip/` and selects the SOC along with any additional SIP specific configs:
```
config SIP_FOO
        bool
        select SOC_BAR
```

bjarki-andreasen-nordic committed:
scripts: list_hardware: add support for sips

Introduce support for SiPs, which are ICs which contain an SoC
potentially along with additional devicetree snippets and
configurations.

The SiP sits between the SoC and the board in HWM2.

SiPs are defined in sip.yml files, which describe the name of the
sip, and which soc it contains.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
cmake: hwm_v2: extend to support SiPs

Generate Kconfig files for SiPs.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
scripts: list_boards: extend to include sip support

Extend board schema, cmake and list_boards.py to support sips. The
sip is used as a board qualifier on the same level as soc.

For example: plank has no socs and one sip. The sip "foo_sip"
contains the soc "bar_soc". The "bar_soc" has two cpuclusters,
"cpuapp" and "cpunet".

The board qualifier for this board would be:
plank/foo_sip/cpuapp/<variant>

The sip does not define cpuclusters, those are "inhereted" from
the soc it contains, the sip is simply defined as:

sips:
  - name: foo_sip
    socs:
      - bar_soc

supporting multiple socs per sip.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
build: add sip support

Add sip support to Kconfig.zephyr and top level CMakeLists.txt along
with an initial top level sip folder.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
sip: correct nrf9161 to the sip schema

The sip nrf9161 is not an soc, its a sip containing an nrf9120 soc.
Change the nrf9161 to be a sip and update boards to select the sip
instead of the now converted soc.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

nordicjm commented on boards/ct/ctcc/board.yml:
```diff
@@ -3,6 +3,9 @@ board:
   full_name: CTHINGS.CO Connectivity Card
   socs:
     - name: nrf52840
+      variants:
+        - name: 'ns'
```
???

bjarki-andreasen replied:
that's a mistake :) PR is very drafty

nordicjm commented on sip/nordic/sip.yml:
```diff
@@ -0,0 +1,3 @@
+sips:
+  - name: nrf9161
+    soc: nrf9120
```
this should be part of the soc.yml file so it's clearly hierarchical though would probably need a rename of the file

bjarki-andreasen replied:
I disagree, just like a board, the sip contains one or more socs, from one or more vendors. The soc is entirely independent of the sip, that's why the sip is defined seperately.

Consider a SiP from vendor `bar`, containing one soc from vendor `foo` and another soc from vendor `baz`, where the sip be defined?

With the schema in this PR, it would be defined in `sip/bar/sip.yml`:
```
sips:
  - name: bar_sip
    socs:
      - foo_soc
      - baz_soc
```

and any SiP specific overlays etc would go in `sip/bar/bar_sip/`.

It is clearly hierarchical, boards contain sips and socs, sips contain socs,

nordicjm replied:
right you are

nordicjm reviewed (commented):
not clear from description if you mean the sip can be omitted entirely if a sip but that should not be allowed, it should work like socs whereby if it is present and there is only one the name can be omitted by the / must still be present

tejlmand reviewed (commented):
Thanks for opening this.

Has been on my wish-list for a long time and looks very close to what I had in mind.

Couple of things not fully clear to me is:
1. Syntax when omitting the SiP or SoC.
2. Overlay files and board conf fragment naming

Ref 1:
> Summarized, after this PR (though not in its current state), the board qualifiers will look like

The examples given only covers when the CPUCluster or variant is not defined, but when defined, and when omitting the SoC when one or more CPU clusters are defined, then it must be replaced by `//`.
Examples:
Single SoC, no CPU clusters `nrf52840dk` --> `nrf52840dk/nrf52840`.
Single SoC, no CPU clusters, variant `nrf52840dk//variant` --> `nrf52840dk/nrf52840/variant.

Single SoC, CPU clusters `nrf5340dk//cpuapp` --> `nrf5340dk/nrf5340/cpuapp`.

Invalid:
`nrf5340dk/cpuapp`
`nrf52840dk/variant`

As the `nrf9161` is a SiP it will always define an SoC, so I assume the syntax will be:
Notice number of `/`.
`nrf9161dk` --> `nrf9161dk/nrf9161/nrf9120`
`nrf9161dk/nrf9161` --> `nrf9161dk/nrf9161/nrf9120`
`nrf9161dk///variant` --> `nrf9161dk/nrf9161/nrf9120/variant`

Illegal:
`nrf9161dk/variant` --> ?
`nrf9161dk//variant` --> ?

bjarki-andreasen-nordic committed:
boards: ct: ctcc: update to use hwm2 sip for nrf9161

The nrf9161 is a sip and has been updated to reflect this. Update
the ctcc board which contains the sip.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
boards: nordic: nrf9161dk: update to use hwm2 sip

The nrf9161 is a sip, containing an nrf9120 soc. Update board
to use the hwm2 sip implementation.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
boards: nordic: nrf9131ek: update to use hwm2 sip

The nrf9131 is a sip, containing an nrf9120 soc. Update board
to use the hwm2 sip implementation.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
boards: nordic: nrf9151dk: update to use hwm2 sip

The nrf9151 is a sip, containing an nrf9120 soc. Update board
to use the hwm2 sip implementation.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
doc: hardware: add sip porting guide

Add sip porting guide and extend the board porting guide to
include sips.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

nashif commented:
Not a fan of this, where does this stop? Will we next introduce SoM, and CoM, PoP, MCM, MCP? 
Why hardware packaging is important for something like Zephyr? If you want to add this complexity to both how hardware is configured and structured, you probably want to do that similar to how we did SoC and the hierarchy we have it using Kconfig. Maybe rename "soc" folder completly to something more generic (hw) and do all of that in there. 

Adding top level folder and the additional layers in HWM IMO is making things more complicated thatn they need to be and naming will become even more cryptic than it is now.

> plank/foo_sip/bar_soc/cpuapp/ns

Please No.

bjarki-andreasen commented:
> Not a fan of this, where does this stop? Will we next introduce SoM, and CoM, PoP, MCM, MCP?

It likely stops at SiPs.

> Why hardware packaging is important for something like Zephyr?

What are boards if not hardware packaging? We care a lot about those right? SoCs are packages of hardware as well.

> If you want to add this complexity to both how hardware is configured and structured, you probably want to do that similar to how we did SoC and the hierarchy we have it using Kconfig. 

That is indeed what this PR does, SiPs sit between boards and SoCs in a clear and accurate hierarchy.

> Maybe rename "soc" folder completly to something more generic (hw) and do all of that in there.

SoCs have a clear definition and position in the HWM hierarchy. This PR exists because treating SiPs as SoCs is not viable. SiPs can contain hardware and SoCs from multiple vendors, just like boards. Attempting to merge SoCs and SiPs will lead to a whole lot of code duplication as SoC code has to be included by multiple SiPs. That's why we have a HWM at in the first place right? to be able to scalably "combine" socs, archs, drivers, and with this PR, SiPs?

> Adding top level folder and the additional layers in HWM IMO is making things more complicated thatn they need to be and naming will become even more cryptic than it is now.

I do not agree the naming is not cryptic. Its clearly and accurately matching the hardware model.

What I do find cryptic is the fact that the `nrf9131`, `nrf9151` and `nrf9161` are all SiPs containing an `nrf9120`. We have been able to abuse the soc schema in this case and treat them like SoCs with Kconfig magic, but that only works until the SiP contains anything that's not a single SoC, from the same vendor.

Naming remains just as complicated for all boards as they are now if they don't have SiPs. Naming only becomes more complicated for the more complicated boards which do contain SiPs.

fkokosinski commented on scripts/schemas/sip-schema.yml:
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) 2025 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+type: map
```
Shouldn't `vendor` field be included in the SiP schema as well?

bjarki-andreasen replied:
Vendor can be added, it is not strictly neccesary outside of filtering so kept it out for now :)

PiotrZierhoffer replied:
From our perspective this would be an important data point - we treat yaml files not only as configuration for filtering but as a source of information about the hardware. 
The same data can be used for documentation generation, so we are strongly for adding this field to the schema as well.

bjarki-andreasen replied:
An earlier version of this PR used the sip/<vendor>/ folder name to determine the vendor, I can add it back :) No need to copy the same information to the sip/<vendor>/sip.yml file I believe :)

fkokosinski replied:
@bjarki-andreasen I wouldn't necessarily say it's copying the same information. We have the exact same case with boards - even though they already are located in `boards/<vendor>`, we still provide the vendor field in `board.yml`. The same holds true for SoCs.
I’d say we should include the vendor field in `sip.yml` because:
* it keeps the format consistent
* we decouple the vendor information from filesystem paths
* it makes data exploration much easier - other than 3rd party tools, it could be for example used by custom Zephyr Sphinx directives for documentation generation and/or other purposes

bjarki-andreasen replied:
hmm

> it keeps the format consistent

sure :)

> we decouple the vendor information from filesystem paths

The folder name is not arbitrary, it has to be the vendor name, so it is copying information

> it makes data exploration much easier - other than 3rd party tools, it could be for example used by custom Zephyr Sphinx directives for documentation generation and/or other purposes

Tools should not, and do not AFAIK read yaml files directly, tools use CMAKE formatted output from the scripts we use to parse the yaml files, like list_hardware.py and list_boards.py so it makes no difference here I believe

henrikbrixandersen commented:
Architecture WG meeting 2025-03-25:
- @bjarki-andreasen gave a presentation of this RFC
- The main concern raised during the comments section was that this modifies the just established HWMv2 with yet another layer, leading to questions regarding the scalability of the current and proposed model.
- Multiple people raised the idea of SiPs (and SoMs) being represented as "sub boards" that can be pulled-in by regular boards (somewhat similar to shields, but "bottom-up" and not user selectable). This would allow for reuse of the existing board infrastructure without adding a new top-level sip/ folder and not modifying the HWMv2 board target scheme, while still being able to reuse the SiP "sub boards" between various board targets using the given SiP.

nordicjm commented:
A previous issue describing a feature that would allow including a board in another board: https://github.com/zephyrproject-rtos/zephyr/issues/55637
